apps/examples/src/examples/speech-bubble/SpeechBubble/helpers.tsx

```diff
@@ -1,211 +1,158 @@
-		new Vec2d(shape.props.w, shape.props.h),
-		new Vec2d(0, shape.props.h),
-	]
+	const [tl, tr, br, bl] = [new Vec2d(0, 0), new Vec2d(w, 0), new Vec2d(w, h), new Vec2d(0, h)]
 
-	if (!intersection) {
-		throw new Error('No intersection')
-	}
+	const offsetH = w / 10
+	const offsetV = h / 10
 
-	const createTailSegments = (orientation: 'horizontal' | 'vertical') => {
-		// Is it a horizontal or vertical line? Which line are we intersecting?
-		return orientation === 'horizontal'
-			? [
-					line === 0
-						? new Vec2d(intersection.x - offset.horizontal, intersection.y)
-						: new Vec2d(intersection.x + offset.horizontal, intersection.y),
-					new Vec2d(handle.x, handle.y),
-					line === 0
-						? new Vec2d(intersection.x + offset.horizontal, intersection.y)
-						: new Vec2d(intersection.x - offset.horizontal, intersection.y),
-			  ]
-			: [
-					line === 1
-						? new Vec2d(intersection.x, intersection.y - offset.vertical)
-						: new Vec2d(intersection.x, intersection.y + offset.vertical),
-					new Vec2d(handle.x, handle.y),
-					line === 1
-						? new Vec2d(intersection.x, intersection.y + offset.vertical)
-						: new Vec2d(intersection.x, intersection.y - offset.vertical),
-			  ]
-	}
+	const { adjustedIntersection, intersectionSegmentIndex } = getHandleIntersectionPoint(shape)
 
-	let modifiedSegments = [...initialSegments]
+	let vertices: Vec2d[]
 
 	// Inject the tail segments into the geometry of the shape
-	switch (line) {
+	switch (intersectionSegmentIndex) {
 		case 0:
-			modifiedSegments.splice(1, 0, ...createTailSegments('horizontal'))
+			// top
+			vertices = [
+				tl,
+				new Vec2d(adjustedIntersection.x - offsetH, adjustedIntersection.y),
+				new Vec2d(handleInShapeSpace.x, handleInShapeSpace.y),
+				new Vec2d(adjustedIntersection.x + offsetH, adjustedIntersection.y),
+				tr,
+				br,
+				bl,
+			]
 			break
 		case 1:
-			modifiedSegments.splice(2, 0, ...createTailSegments('vertical'))
+			// right
+			vertices = [
+				tl,
+				tr,
+				new Vec2d(adjustedIntersection.x, adjustedIntersection.y - offsetV),
+				new Vec2d(handleInShapeSpace.x, handleInShapeSpace.y),
+				new Vec2d(adjustedIntersection.x, adjustedIntersection.y + offsetV),
+				br,
+				bl,
+			]
 			break
 		case 2:
-			modifiedSegments.splice(3, 0, ...createTailSegments('horizontal'))
+			// bottom
+			vertices = [
+				tl,
+				tr,
+				br,
+				new Vec2d(adjustedIntersection.x + offsetH, adjustedIntersection.y),
+				new Vec2d(handleInShapeSpace.x, handleInShapeSpace.y),
+				new Vec2d(adjustedIntersection.x - offsetH, adjustedIntersection.y),
+				bl,
+			]
 			break
 		case 3:
-			modifiedSegments = [...modifiedSegments, ...createTailSegments('vertical')]
+			// left
+			vertices = [
+				tl,
+				tr,
+				br,
+				bl,
+				new Vec2d(adjustedIntersection.x, adjustedIntersection.y + offsetV),
+				new Vec2d(handleInShapeSpace.x, handleInShapeSpace.y),
+				new Vec2d(adjustedIntersection.x, adjustedIntersection.y - offsetV),
+			]
```

Comment:
This was another place where I had to squint pretty hard to get what was going on. This is IMO a good spend of repetition in order to communicate (visually, even!) what's going on.